### PR TITLE
Data Connector for Armorblox Solution

### PR DESCRIPTION
**Pull request by ujayant**

### CODE REVIEW

**Comment by ujayant** [comment]
Looks like the `KqlValidations` are failing even when run for the master branch of the Sentinel repo. Not sure if these failures have got anything to do with this PR.

File: Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                        "fillWith": ["PrimaryKey"],
+                        "label": "Primary Key"
+                    },
+                    "type": "CopyableLabel"
+                }
+            ]
+        },
+        {
+            "title": "Option 1 - Azure Resource Manager (ARM) Template",
+            "description": "Use this method for automated deployment of the Armorblox data connector using an ARM Tempate.\n\n1. Click the **Deploy to Azure** button below. \n\n\t[![Deploy To Azure](https://aka.ms/deploytoazurebutton)](https://raw.githubusercontent.com/armorblox/Azure-Sentinel/master/Solutions/Armorblox/Data%20Connectors/azuredeploy_Armorblox_API_FunctionApp.json)\n2. Select the preferred **Subscription**, **Resource Group** and **Location**. \n> **NOTE:** Within the same resource group, you can't mix Windows and Linux apps in the same region. Select existing resource group without Windows apps in it or create new resource group.\n3. Enter the **ArmorbloxAPIToken**, **ArmorbloxInstanceURL** OR **ArmorbloxInstanceName**, and deploy. \n4. Mark the checkbox labeled **I agree to the terms and conditions stated above**. \n5. Click **Purchase** to deploy."
```
Comment (NikTripathi) [inline Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json:110]: Fix typo Tempate as Template.

File: Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
...Functions** area, choose the **Deploy to function app** button.\nIf you aren't already signed in, choose the Azure icon in the Activity bar, then in the **Azure: Functions** area, choose **Sign in to Azure**\nIf you're already signed in, go to the next step.\n5. Provide the following information at the prompts:\n\n\ta. **Select folder:** Choose a folder from your workspace or browse to one that contains your function app.\n\n\tb. **Select Subscription:** Choose the subscription to use.\n\n\tc. Select **Create new Function App in Azure** (Don't choose the Advanced option)\n\n\td. **Enter a globally unique name for the function app:** Type a name that is valid in a URL path. The name you type is validated to make sure that it's unique in Azure Functions. (e.g. Armorblox).\n\n\te. **Select a runtime:** Choose Python 3.8.\n\n\tf. Select a location for new resources. For better performance and lower costs choose the same [region](https://azure.microsoft.com/regions/) where Azure Sentinel is located.\n\n6. Deployment will begin. A notification is displayed after your function app is created and the deployment package is applied.\n7. Go to Azure Portal for the Function App configuration."
```
Comment (NikTripathi) [inline Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json:118]: Choose the subscription to use.\n\n\tc. (e.g. Armorblox).\n\n\te , Choose Python 3.8.\n\n\tf.  Why a c,e and f is at the end?

File: Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json (hunk 1 above)
Comment (NikTripathi) [inline Solutions/Armorblox/Data Connectors/Armorblox_API_FunctionApp.json:110]: Please update deployment link as **https://aka.ms/sentinel-armorblox-azuredeploy**

File: Solutions/Armorblox/Data Connectors/azuredeploy_Armorblox_API_FunctionApp.json [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+                        "WEBSITE_CONTENTSHARE": "[toLower(variables('FunctionName'))]",
+                        "AzureWebJobsStorage": "[concat('DefaultEndpointsProtocol=https;AccountName=', toLower(variables('FunctionName')),';AccountKey=',listKeys(resourceId('Microsoft.Storage/storageAccounts', toLower(variables('FunctionName'))), '2019-06-01').keys[0].value, ';EndpointSuffix=',toLower(variables('StorageSuffix')))]",
+                        "WorkspaceID": "[parameters('WorkspaceID')]",
+                        "WorkspaceKey": "[parameters('WorkspaceKey')]",
+                        "ArmorbloxAPIToken": "[parameters('ArmorbloxAPIToken')]",
+                        "ArmorbloxInstanceName": "[parameters('ArmorbloxInstanceName')]",
+                        "ArmorbloxInstanceURL": "[parameters('ArmorbloxInstanceURL')]",
+                        "LogAnalyticsUri": "[variables('LogAnalyticsUri')]",
+                        "WEBSITE_RUN_FROM_PACKAGE": "https://raw.githubusercontent.com/armorblox/Azure-Sentinel/master/Solutions/Armorblox/Data%20Connectors/ArmorbloxAzureFunction.zip"
```
Comment (NikTripathi) [inline Solutions/Armorblox/Data Connectors/azuredeploy_Armorblox_API_FunctionApp.json:180]: Please update WEBSITE_RUN_FROM_PACKAGE link as **https://aka.ms/sentinel-armorblox-functionapp**

**Review by NikTripathi — CHANGES_REQUESTED** [review]

**Comment by NikTripathi** [comment]
Please provide sample data in JSON format.

**Comment by NikTripathi** [comment]
> Please provide sample data in JSON format.
We have used .csv format so no need to update JSON format. Thank You.

**Comment by urajat** [comment]
@NikTripathi Thanks for the review! All review comments have been addressed.

**Comment by NikTripathi** [comment]
During the deployment function app is loading. But function is not loading in portal. I think it is a problem because of the structure of zip file. Please have a look on the zip file structure of [CiscoUmbrella](https://github.com/Azure/Azure-Sentinel/tree/master/DataConnectors/CiscoUmbrella). Let me know if something needs to done by my end.

**Review by NikTripathi — CHANGES_REQUESTED** [review]
Data connector deployment is successful but function name is still not loading. See image for reference.  image

**Comment by ujayant** [comment]
@NikTripathi I'm facing the same issue, and I'm trying to figure it out. It worked for the first time, but subsequent deployments have this issue. 

I faced the same issue even when I just deployed the function app for testing. Changing the name of the function app fixed the issue for me. Any ideas on this?

**Comment by NikTripathi** [comment]
> @NikTripathi I'm facing the same issue, and I'm trying to figure it out. It worked for the first time, but subsequent deployments have this issue.
> 
> I faced the same issue even when I just deployed the function app for testing. Changing the name of the function app fixed the issue for me. Any ideas on this?

I changed the function name and redeployed it. Still facing the same issue. Function deploys successfully but function is not loading.

**Comment by urajat** [comment]
@NikTripathi We tried the steps under Manual Deployment of Azure Functions section and that seems to work correctly - we are able to see the function app files load in the portal after deploying via VSCode. Only when we try the ARM template route using the Deploy To Azure button, it seems to show the behaviour that you are seeing with the deployment being successful, but function not loading. We are unable to spot any material differences between the Cisco one and ours. We also cross-checked against a couple of others like SentinelOne for e.g.

**Comment by ujayant** [comment]
@NikTripathi It looks like there was an extra top-level folder in the zip file. I was able to test it through template-based deployment, and it worked properly. Let me know if you find anything else.

**Comment by v-jayakal** [comment]
@ujayant - Function gets deployed successfully from template. Invocation logs show error that 'requests' module is missing. I am checking whether there is version issue at my end. Is it running good at your end without error? I will give you update about the version issue at my end by EOD. Please let us know about function invocation result  at your end. Thanks.

**Comment by NikTripathi** [comment]
Provided zip package doesn't have .python-package folder. I added the python package and tested it. Can you please update the old zip package with the attached zip package.
[ArmorbloxAzureFunction.zip](https://github.com/Azure/Azure-Sentinel/files/7284911/ArmorbloxAzureFunction.zip)

**Comment by ujayant** [comment]
@NikTripathi I've updated and tested this end-to-end. Everything is working fine.

**Comment by v-jayakal** [comment]
@ujayant - Its working now, thanks.

**Review by v-jayakal — APPROVED** [review]

**Pull request merged**